For #8893 - Add a new partial bind method to TabViewHolder

Mugurell opened the pull request:
This new method allows temporarily suppressing the 'selected tab' decoration
and then re-enable the default behavior.

The result of using this on Fenix:
![selectAndCheck](https://user-images.githubusercontent.com/11428869/99394529-b38f4700-28e7-11eb-8784-8d2da2e6df98.gif)
[video](https://drive.google.com/file/d/1HAAG-Bn1qpbOlp0vHzH2QjZvnLzN2g4n/view?usp=sharing)

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks.
- [x] **Tests**: This PR includes thorough tests.
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md).
- [x] **Accessibility**: The code in this PR does not include any a11y user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

jonalmeida commented on components/browser/tabstray/src/main/java/mozilla/components/browser/tabstray/TabViewHolder.kt:
```diff
@@ -39,6 +40,15 @@ abstract class TabViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+     */
+    open fun updateSelectedTabIndicator(showAsSelected: Boolean) {
+        // no-op.
+        // Not an abstract fun since not all clients of this library might be interested in this functionality.
```
nit: Maybe throw an `UnsupportedOperationException` for the default behaviour so a consumer isn't confused if this is called and nothing happens.

Mugurell replied:
I like this. Thanks!

jonalmeida commented on components/browser/tabstray/src/main/java/mozilla/components/browser/tabstray/TabsAdapter.kt:
```diff
@@ -65,4 +86,20 @@ open class TabsAdapter(
+         *
+         * Signals that the currently selected tab should be highlighted. This is the default behavior.
+         */
+        const val PAYLOAD_HIGHLIGHT_SELECTED_ITEM = "Should highlight the current tab"
```
Should these just be integers? That could suffice with indicating the type, maybe a resource generated value in the `ids.xml` so we don't have to keep count of integer increments.

```suggestion
        const val PAYLOAD_HIGHLIGHT_SELECTED_ITEM = R.id.payload_highlight_selected_item
```

Mugurell replied:
Sure! Thanks!

jonalmeida reviewed (approved):
LGTM, left a nit and one suggestion to avoid an extra string allocation for each ViewHolder. 🙂

Mugurell commented:
Thank you for helping with this @jonalmeida!
Codrut saw [an interesting issue](https://github.com/mozilla-mobile/fenix/pull/16370#pullrequestreview-529049145) while reviewing the Fenix patch, if the items are not binded the partial bind will not work for them so if the selected item is scrolled away from View we would not know to remove the "selected indicator from it", something [the Android documentation do mention](https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.Adapter#notifyItemChanged(int,%20java.lang.Object)):
> Adapter should not assume that the payload will always be passed to onBindViewHolder(), e.g. when the view is not attached, the payload will be simply dropped.

Hoping we can still use this way of partial binding and not have to fully rebind all items, will be looking into what's still needed.

Mugurell commented:
The new changes addresses the review comments 
and add a new `TabsAdapter#isTabSelected(..)` method that complements the previous patch for partial binding and allows now for clients like Fenix to override this when doing full binds for items not shown at the time of the partial bind
allowing clients to control whether to apply or not the "selected item" decorator.

jonalmeida commented on components/concept/tabstray/src/main/java/mozilla/components/concept/tabstray/TabsTray.kt:
```diff
@@ -63,6 +63,11 @@ interface TabsTray : Observable<TabsTray.Observer> {
+    /**
+     * Called when binding a new item to get if it should be shown as selected or not.
+     */
+    fun isTabSelected(tabs: Tabs, position: Int): Boolean
```
```suggestion
    fun isTabSelected(tabs: Tabs, position: Int): Boolean = tabs.selectedIndex == position
```

What do you think about having a default implementation that do the basic check for you? I'm not fully convinced I like default impls like this on an API, but I thought I'd throw the idea out there either way. 😄

Mugurell replied:
Thanks!
I also thought about this but saw that other methods from this interface also have simple implementations but not as default.
So mainly just to be consistent with them I put the implementation in `TabsAdapter`.
This simple check is the only reason I went with this Fenix - AC - Fenix loop. The list of tabs is in AC, the behavior for if to show a tab selected or not is in Fenix.

I like the default implementation as that would make this a non-breaking change, what I initially intended, but as you've said maybe there shouldn't be any logic at this level.
I would prefer not to be the first that starts polluting this interfaces 😂, will add the breaking change mention if it's ok with you.

jonalmeida reviewed (approved):
Sorry about the slow review! Left one question and a changelog nit.

Mugurell commented:
Thank you for helping with this Jon.
Since this would be breaking for Fenix I'll I'll wait for having a review there also before merging.

Mugurell commented on docs/changelog.md:
```diff
@@ -12,6 +12,9 @@ permalink: /changelog/
 * [Configuration](https://github.com/mozilla-mobile/android-components/blob/master/.config.yml)
 
+* **browser-tabstray**
+  * Support temporary hiding if a tab is selected. For this you should use partial bindings with PAYLOAD_(DONT_)HIGHLIGHT_SELECTED_ITEM; override TabsAdapter#isTabSelected for new item bindings; override TabViewHolder#updateSelectedTabIndicator.
```
Done. Thanks!
